plugin/pulpcore/plugin/download/asyncio/base.py

```diff
@@ -0,0 +1,153 @@
+
+        Args:
+            data (bytes): The data to have its size and digest values recorded.
+        """
+        for algorithm in self._digests.values():
+            algorithm.update(data)
+        self._size = self._size + len(data)
+
+    @property
+    def artifact_attributes(self):
+        """
+        A property that returns a dictionary with size and digest information. The keys of this
+        dictionary correspond with :class:`pulpcore.plugin.models.Artifact` fields.
+        """
+        attributes = {'size': self._size}
+        for algorithm in Artifact.DIGEST_FIELDS:
+            attributes[algorithm] = self._digests[algorithm].hexdigest()
+        return attributes
+
+    def validate_digests(self):
+        """
+        Validate all digests validate if ``expected_digests`` is set
+
+        Raises:
+            :class:`~pulpcore.plugin.download.asyncio.DigestValidationError`: When any of the
+                ``expected_digest`` values don't match the digest of the data passed to
+                :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data`.
+        """
+        if self.expected_digests:
+            for algorithm, expected_digest in self.expected_digests.items():
+                if expected_digest != self._digests[algorithm].hexdigest():
+                    raise DigestValidationError()
+
+    def validate_size(self):
+        """
+        Validate the size if ``expected_size`` is set
+
+        Raises:
+            :class:`~pulpcore.plugin.download.asyncio.SizeValidationError`: When the
+                ``expected_size`` value doesn't match the size of the data passed to
+                :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data`.
+        """
+        if self.expected_size:
+            if self._size != self.expected_size:
+                raise SizeValidationError()
+
+    async def run(self):
+        """
+        Run the downloader.
+
+        This is a coroutine that asyncio can schedule to complete downloading. This is required to
+        be implemented by subclasses.
+
+        It is expected that the subclass implementation call
+        :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.validate_size` and
+        :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.validate_digests` after the last
+        call to :meth:`~pulpcore.plugin.download.asyncio.BaseDownloader.handle_data` which will
+        validate the data.
+
+        It is also expected that the subclass implementation return a
+        :class:`~pulpcore.plugin.download.asyncio.DownloadResult` object. The
+        ``artifact_attributes`` value of the
+        :class:`~pulpcore.plugin.download.asyncio.DownloadResult` is usually set to the
+        :attr:`~pulpcore.plugin.download.asyncio.BaseDownloader.artifact_attributes` property value.
+
+        Returns:
+            :class:`~pulpcore.plugin.download.asyncio.DownloadResult`
```

Comment:
This method actually returns a `coroutine` not a `DownloadResult` and It's the coroutine that actually returns a `DownloadResult`, right?  If so, I'm not sure how to document that but this seems misleading.

Reply:
I observed the same thing. I'm following what the `aiohttp` library doe,s and they have the return value of coroutines documented as the final return value emitted from the coroutine itself. Of the two choices only this one lets me express what the ultimate result will be. Otherwise, I'm not sure how to declare the result of the coroutine if I use the return value to declare that it is a coroutine. With that thinking I'm going to leave as is. I am going to check the three `run()` implementations to make sure their return values are consistent also.

Reply:
Oh yeah the other subclasses only refer to this docstring.